[IDEA] When debugging at a point this null pointer exception thrown from Idea

**Description:**

When debugging on a line where there is a record to JSON conversion following error thrown from IDEA.

```log
java.lang.NullPointerException
	at io.ballerina.plugins.idea.debugger.BallerinaXValue.getPresentation(BallerinaXValue.java:99)
	at io.ballerina.plugins.idea.debugger.BallerinaXValue.computePresentation(BallerinaXValue.java:74)
	at com.intellij.xdebugger.impl.ui.tree.nodes.XValueNodeImpl.<init>(XValueNodeImpl.java:57)
	at com.intellij.xdebugger.impl.ui.tree.nodes.XValueContainerNode.lambda$addChildren$0(XValueContainerNode.java:89)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.doRun(LaterInvocator.java:435)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:419)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:403)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:729)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:678)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:373)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

I notice the following issue also.

```bash
java.lang.AssertionError: Already disposed: Project (Disposed) bal_project
	at com.intellij.openapi.components.impl.ComponentManagerImpl.lambda$throwAlreadyDisposed$1(ComponentManagerImpl.java:245)
	at com.intellij.openapi.application.ReadAction.lambda$run$1(ReadAction.java:53)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:974)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:57)
	at com.intellij.openapi.application.ReadAction.run(ReadAction.java:53)
	at com.intellij.openapi.components.impl.ComponentManagerImpl.throwAlreadyDisposed(ComponentManagerImpl.java:243)
	at com.intellij.openapi.components.impl.ComponentManagerImpl.getPicoContainer(ComponentManagerImpl.java:236)
	at com.intellij.openapi.components.impl.ComponentManagerImpl.getComponent(ComponentManagerImpl.java:146)
	at com.intellij.openapi.fileEditor.FileEditorManager.getInstance(FileEditorManager.java:21)
	at org.wso2.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.connect(LanguageServerWrapper.java:637)
	at org.wso2.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.reconnect(LanguageServerWrapper.java:591)
	at org.wso2.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.crashed(LanguageServerWrapper.java:557)
	at org.wso2.lsp4intellij.editor.EditorEventManager.requestAndShowDoc(EditorEventManager.java:822)
	at org.wso2.lsp4intellij.editor.EditorEventManager.access$200(EditorEventManager.java:164)
	at org.wso2.lsp4intellij.editor.EditorEventManager$1.run(EditorEventManager.java:948)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
```

--- a/tool-plugins/intellij/gradle.properties
+++ b/tool-plugins/intellij/gradle.properties
@@ -14,5 +14,5 @@
 #  limitations under the License.
 #
 ballerinaPluginName=ballerina-intellij-idea-plugin
-ballerinaPluginVersion=1.0.0-alpha-2
+ballerinaPluginVersion=1.0.0-alpha-3
 ideaVersion=2019.2

--- a/tool-plugins/intellij/src/main/gen/io/ballerina/plugins/idea/parser/BallerinaParser.java
+++ b/tool-plugins/intellij/src/main/gen/io/ballerina/plugins/idea/parser/BallerinaParser.java
@@ -5047,10 +5047,10 @@ public class BallerinaParser implements PsiParser, LightPsiParser {
     boolean r, p;
     Marker m = enter_section_(b, l, _NONE_, SERVICE_DEFINITION, null);
     r = consumeToken(b, SERVICE);
-    p = r; // pin = 1
-    r = r && report_error_(b, ServiceDefinition_1(b, l + 1));
-    r = p && report_error_(b, consumeToken(b, ON)) && r;
-    r = p && report_error_(b, ExpressionList(b, l + 1)) && r;
+    r = r && ServiceDefinition_1(b, l + 1);
+    r = r && consumeToken(b, ON);
+    p = r; // pin = 3
+    r = r && report_error_(b, ExpressionList(b, l + 1));
     r = p && ServiceBody(b, l + 1) && r;
     exit_section_(b, l, m, r, p, null);
     return r || p;

--- a/tool-plugins/intellij/src/main/gen/io/ballerina/plugins/idea/psi/BallerinaServiceDefinition.java
+++ b/tool-plugins/intellij/src/main/gen/io/ballerina/plugins/idea/psi/BallerinaServiceDefinition.java
@@ -33,7 +33,7 @@ public interface BallerinaServiceDefinition extends BallerinaTopLevelDefinition 
   @Nullable
   PsiElement getIdentifier();
 
-  @Nullable
+  @NotNull
   PsiElement getOn();
 
   @NotNull

--- a/tool-plugins/intellij/src/main/gen/io/ballerina/plugins/idea/psi/impl/BallerinaServiceDefinitionImpl.java
+++ b/tool-plugins/intellij/src/main/gen/io/ballerina/plugins/idea/psi/impl/BallerinaServiceDefinitionImpl.java
@@ -61,9 +61,9 @@ public class BallerinaServiceDefinitionImpl extends ASTWrapperPsiElement impleme
   }
 
   @Override
-  @Nullable
+  @NotNull
   public PsiElement getOn() {
-    return findChildByType(ON);
+    return findNotNullChildByType(ON);
   }
 
   @Override

--- a/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/debugger/BallerinaDebugProcess.java
+++ b/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/debugger/BallerinaDebugProcess.java
@@ -133,10 +133,10 @@ public class BallerinaDebugProcess extends XDebugProcess {
     public void sessionInitialized() {
         final int[] retryAttempt = {0};
         ApplicationManager.getApplication().executeOnPooledThread(() -> {
-
+            getSession().getConsoleView().print("Connecting to the debug server...\n",
+                    ConsoleViewContentType.SYSTEM_OUTPUT);
             // If already connected with the debug server, tries to set breakpoints and attach with the remote jvm.
             if (dapClientConnector.isConnected()) {
-
                 LOGGER.debug("Connection is already created.");
                 isConnected = true;
                 startDebugSession();
@@ -505,6 +505,7 @@ public class BallerinaDebugProcess extends XDebugProcess {
                     // Sends attach request to the debug server.
                     LOGGER.debug("Sending Attach command.");
                     dapClientConnector.attachToServer();
+                    getSession().getConsoleView().print("Compiling...\n", ConsoleViewContentType.SYSTEM_OUTPUT);
                 }
             });
         }

--- a/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/debugger/BallerinaDebugProcess.java
+++ b/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/debugger/BallerinaDebugProcess.java
@@ -170,9 +170,6 @@ public class BallerinaDebugProcess extends XDebugProcess {
                     startDebugSession();
                     break;
                 }
-                if (isRemoteDebugMode) {
-                    break;
-                }
             }
             if (!dapClientConnector.isConnected()) {
                 getSession().getConsoleView().print("Connection to debug server at " +
@@ -256,24 +253,26 @@ public class BallerinaDebugProcess extends XDebugProcess {
     @Override
     public void stop() {
         // If we don't call this using the executeOnPooledThread(), the UI will hang until the debug server is stopped.
-        if (isConnected) {
-            ApplicationManager.getApplication().executeOnPooledThread(() -> {
-                try {
-                    dapClientConnector.disconnectFromServer();
-                    getSession().getConsoleView().print("Disconnected Successfully from the debug server.\n",
-                            ConsoleViewContentType.SYSTEM_OUTPUT);
-                } catch (Exception e) {
-                    getSession().getConsoleView().print("Disconnected Exceptionally from the debug server.\n",
-                            ConsoleViewContentType.SYSTEM_OUTPUT);
-                } finally {
-                    XDebugSession session = getSession();
-                    if (session != null) {
-                        session.stop();
-                    }
-                    isConnected = false;
+        ApplicationManager.getApplication().invokeLater(() -> {
+            if (!isConnected) {
+                return;
+            }
+            try {
+                dapClientConnector.disconnectFromServer();
+                getSession().getConsoleView().print("Disconnected Successfully from the debug server.\n",
+                        ConsoleViewContentType.SYSTEM_OUTPUT);
+            } catch (Exception e) {
+                getSession().getConsoleView().print("Disconnected Exceptionally from the debug server.\n",
+                        ConsoleViewContentType.SYSTEM_OUTPUT);
+            } finally {
+                XDebugSession session = getSession();
+                if (session != null) {
+                    session.stop();
                 }
-            });
-        }
+                isConnected = false;
+            }
+        });
+
     }
 
     @Nullable

--- a/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/debugger/BallerinaXValue.java
+++ b/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/debugger/BallerinaXValue.java
@@ -106,8 +106,8 @@ public class BallerinaXValue extends XNamedValue {
 
     @NotNull
     private XValuePresentation getPresentation() {
-        String value = variable.getValue();
-        String type = variable.getType();
+        String value = variable.getValue() != null ? variable.getValue() : "";
+        String type = variable.getType() != null ? variable.getType() : "";
 
         if (type.equalsIgnoreCase(BallerinaXValueType.STRING.getValue())) {
             return new XValuePresentation() {

--- a/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/grammar/Ballerina.bnf
+++ b/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/grammar/Ballerina.bnf
@@ -154,7 +154,8 @@ Definition ::= ( TypeDefinition
 
 private TopLevelDefinitionRecover ::= !(MARKDOWN_DOCUMENTATION_LINE_START|PARAMETER_DOCUMENTATION_START|RETURN_PARAMETER_DOCUMENTATION_START|'@'|external|remote|client|abstract|public|type|typedesc|service|listener|function|enum|annotation|int|float|decimal|boolean|string|byte|map|xml|xmlns|json|table|any|stream|object|record|channel|const|final|var|future|identifier|'{')
 
-ServiceDefinition ::= service identifier? on ExpressionList ServiceBody {pin=1}
+// cannot set pin to "service" keyword because there can be variable definitions with service constructors.
+ServiceDefinition ::= service identifier? on ExpressionList ServiceBody {pin=3}
 ServiceBody ::= LEFT_BRACE ObjectFunctionDefinition* RIGHT_BRACE
 
 // Pin to make sure callable unit is identified correctly

--- a/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/BallerinaConsoleFolding.java
+++ b/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/BallerinaConsoleFolding.java
@@ -35,12 +35,18 @@ public class BallerinaConsoleFolding extends ConsoleFolding {
     @Override
     public boolean shouldFoldLine(@NotNull String line) {
         Matcher matcher = PATTERN.matcher(line);
-        return matcher.find();
+        return matcher.find() || isDebugLog(line);
     }
 
     @Nullable
     @Override
     public String getPlaceholderText(@NotNull List<String> lines) {
         return "<...>";
+    }
+
+    // Todo - Remove after moving debug adapter into the same JVM instance.
+    private boolean isDebugLog(@NotNull String line) {
+        return line.contains("Please start the remote debugging client to continue")
+                || line.contains("Listening for transport dt_socket");
     }
 }

--- a/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java
+++ b/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java
@@ -97,8 +97,7 @@ public class BallerinaApplicationRunningState extends BallerinaRunningState<Ball
                 project.getBasePath());
         // if no ballerina project is found.
         if (projectRoot.isEmpty()) {
-            projectRoot = projectPath;
-            filePath = Paths.get(file.getVirtualFile().getPath()).toString().replace(projectRoot, "").substring(1);
+            filePath = Paths.get(file.getVirtualFile().getPath()).toString();
         } else {
             String relativeFilePath =
                     Paths.get(file.getVirtualFile().getPath()).toString().replace(projectRoot, "").substring(1);

--- a/tool-plugins/intellij/src/main/resources/META-INF/plugin.xml
+++ b/tool-plugins/intellij/src/main/resources/META-INF/plugin.xml
@@ -1,7 +1,7 @@
 <idea-plugin>
     <id>io.ballerina</id>
     <name>Ballerina</name>
-    <version>1.0.0-alpha-2</version>
+    <version>1.0.0-alpha-3</version>
     <vendor email="ballerinalang@wso2.com" url="https://ballerina.io/">BallerinaLang</vendor>
 
     <description><![CDATA[

